<?xml version="1.0"?>
<!--
MIT License
Copyright (c) 2024 Naoki Takahashi
-->
<robot xmlns:xacro="http://www.ros.org/wiki/xacro">
  <xacro:macro name="cybergear_servo" params="name parent *origin">
    <joint name="${name}_FR1_base_joint" type="fixed">
      <origin rpy="${-pi/2} 0 0" xyz="0.0 0 0.1"/>
      <parent link="${parent}"/>
      <child link="${name}_FR1_link"/>
    </joint>
    <link name="${name}_FR1_link">
      <visual>
        <geometry>
          <mesh filename="package://quadro/meshes/cybergear_body.stl" scale="0.001 0.001 0.001"/>
        </geometry>
      </visual>
      <collision>
        <origin rpy="0 0 0" xyz="0 0 0.008"/>
        <geometry>
          <cylinder length="0.035" radius="0.042"/>
        </geometry>
      </collision>

    </link>
    <joint name="FR1" type="revolute">
      <origin rpy="0 0 0" xyz="0 0 0.028"/>
      <parent link="${name}_FR1_link"/>
      <child link="${name}_FR1_table_link"/>
      <axis xyz="0 0 1"/>
      <limit effort="12" velocity="30" lower="-12.565" upper="12.565"/>
    </joint>
    <link name="${name}_FR1_table_link">
      <visual>
        <origin rpy="0 0 0" xyz="0 0 -0.028"/>
        <geometry>
          <mesh filename="package://quadro/meshes/cybergear_table.stl" scale="0.001 0.001 0.001"/>
        </geometry>
      </visual>
      <collision>
        <origin rpy="0 0 0" xyz="0 0 -0.005"/>
        <geometry>
          <cylinder length="0.01" radius="0.016"/>
        </geometry>
      </collision>

    </link>

    <!-- SECOND LINK -->

    <joint name="${name}_FR2_base_joint" type="fixed">
      <origin rpy="0 0 0" xyz="0.1 0 0.0"/>
      <parent link="${parent}"/>
      <child link="${name}_FR2_link"/>
    </joint>
    <link name="${name}_FR2_link">
      <visual>
        <geometry>
          <mesh filename="package://quadro/meshes/cybergear_body.stl" scale="0.001 0.001 0.001"/>
        </geometry>
      </visual>
      <collision>
        <origin rpy="0 0 0" xyz="0 0 0.008"/>
        <geometry>
          <cylinder length="0.035" radius="0.042"/>
        </geometry>
      </collision>

    </link>
    <joint name="FR2" type="revolute">
      <origin rpy="0 0 0" xyz="0 0 0.028"/>
      <parent link="${name}_FR2_link"/>
      <child link="${name}_FR2_table_link"/>
      <axis xyz="0 0 1"/>
      <limit effort="12" velocity="30" lower="-12.565" upper="12.565"/>
    </joint>
    <link name="${name}_FR2_table_link">
      <visual>
        <origin rpy="0 0 0" xyz="0 0 -0.028"/>
        <geometry>
          <mesh filename="package://quadro/meshes/cybergear_table.stl" scale="0.001 0.001 0.001"/>
        </geometry>
      </visual>
      <collision>
        <origin rpy="0 0 0" xyz="0 0 -0.005"/>
        <geometry>
          <cylinder length="0.01" radius="0.016"/>
        </geometry>
      </collision>

    </link>

    <ros2_control name="CybergearActuators" type="actuator">
    <hardware>
      <plugin>quadro/CybergearActuator</plugin>
      <param name="can_interface">can0</param>
      <param name="timeout_sec">3.0</param>
      <param name="use_bus_time">false</param>
      <param name="interval_sec">0.5</param>
      <param name="device_id">127</param>
      <param name="primary_id">0</param>
      <param name="max_position">12.56637061</param>
      <param name="min_position">-12.56637061</param>
      <param name="max_velocity">30.0</param>
      <param name="min_velocity">-30.0</param>
      <param name="max_effort">12.0</param>
      <param name="min_effort">-12.0</param>
      <param name="max_gain_kp">500.0</param>
      <param name="min_gain_kp">0.0</param>
      <param name="max_gain_kd">5.0</param>
      <param name="min_gain_kd">0.0</param>
      <param name="max_current">23.0</param>
      <param name="min_current">-23.0</param>
      <param name="temperature_scale">0.1</param>
    </hardware>

    <joint name="FR1">
      <command_interface name="position">
      </command_interface>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
    </joint>
    <joint name="FR2">
      <command_interface name="position">
      </command_interface>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
    </joint>
  </ros2_control>
  </xacro:macro>
</robot>
